src/Microsoft.DotNet.Archive/IndexedArchive.cs

```diff
@@ -0,0 +1,583 @@
+                                string destinationPath = Path.Combine(outputDirectory, zipRelativePath);
+
+                                // operations on a zip file will be sequential
+                                ZipOperation currentZipOperation;
+
+                                if (!zipOperations.TryGetValue(destinationPath, out currentZipOperation))
+                                {
+                                    extractOperations.Add(currentZipOperation = new ZipOperation(destinationPath));
+                                    zipOperations.Add(destinationPath, currentZipOperation);
+                                }
+                                currentZipOperation.AddEntry(zipEntryName, extractSource);
+                            }
+                            else
+                            {
+                                string destinationPath = Path.Combine(outputDirectory, target);
+                                extractOperations.Add(new CopyOperation(extractSource, destinationPath));
+                            }
+                        }
+                    }
+
+                    int opsExecuted = 0;
+                    // execute all operations
+                    //foreach(var extractOperation in extractOperations)
+                    extractOperations.AsParallel().ForAll(extractOperation =>
+                    {
+                        extractOperation.DoOperation();
+                        progress.Report("Expanding", Interlocked.Increment(ref opsExecuted), extractOperations.Count);
+                    });
+                }
+            }
+        }
+
+        public void AddExternalDirectory(string externalDirectory)
+        {
+            CheckDisposed();
+            foreach (var externalFile in Directory.EnumerateFiles(externalDirectory, "*", SearchOption.AllDirectories))
+            {
+                AddExternalFile(externalFile);
+            }
+        }
+
+        public void AddExternalFile(string externalFile)
+        {
+            CheckDisposed();
+            using (var fs = File.OpenRead(externalFile))
+            {
+                string hash = GetHash(fs); 
+                // $ prefix indicates that the file is not in the archive and path is relative to an external directory
+                archiveFiles[hash] = new ArchiveFileInfo(null, "$" + hash , hash);
+                externalFiles[hash] = externalFile;
+            }
+        }
+        public void AddDirectory(string sourceDirectory, IProgress<ProgressReport> progress, string destinationDirectory = null)
+        {
+            var sourceFiles = Directory.EnumerateFiles(sourceDirectory, "*", SearchOption.AllDirectories).ToArray();
+            int filesAdded = 0;
+            sourceFiles.AsParallel().ForAll(sourceFile =>
+                {
+                    string destinationPath = sourceFile.Substring(sourceDirectory.Length + 1);
```

Comment:
It's not a file name.  Its a trimmed path under the sourceDirectory.  It may have subfolders.  Perhaps destinationPartialPath?